Use ReactDOM.hydrate() for hydrating a SSR component if available

theJoeBiz opened the pull request:
Looks like this was originally addressed in #1013, but was closed by accident and never reopened. My project is using React 16 and I can confirm that this patch prevents the deprecation warning.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/shakacode/react_on_rails/1028)
<!-- Reviewable:end -->

justin808 commented:
@theJoeBiz Any way to test or verify this for both older React and new react?

theJoeBiz commented:
@justin808 I don't think so. There aren't any existing tests for `clientStartup.js` and I've never had to deal with tests for two separate versions of react. That being said, the change made here is fairly transparent and visibly falls through to the old behavior for any version (older than 16) where `ReactDOM.hydrate` is undefined.

theJoeBiz commented:
Also, only one test needs to be updated for a react 16 upgrade
https://github.com/theJoeBiz/react_on_rails/commit/468338c1ab0964e872efe45d5cd4e994b1515ee8#diff-85a5145721dbc772cfb6de2e6102717c

justin808 commented:
@alexfedoseev can you please confirm this React syntax looks good.

alexfedoseev commented on node_package/src/clientStartup.js:
```diff
@@ -128,7 +128,17 @@ function render(el, railsContext) {
+          ReactDOM.hydrate &&
+          domNode &&
+          domNode.hasAttribute('data-reactroot')
+        );
```
Looks like you mix things here you shouldn't mix. E.g. `!!ReactDOM.hydrate` is enough for `shouldHydrate` decision, but if something is wrong w/ dom node then it's out of scope `ReactDOM.hydrate` vs `ReactDOM.render` problem.

Also, https://github.com/facebook/react/issues/10971

theJoeBiz commented:
@alexfedoseev Good call, updated to just check for truthiness of ReactDOM.hydrate.

justin808 commented:
@alexfedoseev let me know if I should merge.

theJoeBiz commented:
We just moved out app into production with this warning, which is fine, but I was just revisiting this on a page that has a combination of server/client rendered components and found a flaw. Turns out, the `data-reactroot` check I removed in a previous iteration of this was there to check to see if the component was server rendered. In react 16, `ReactDOM.hydrate` is only supposed to be used on components that were previously rendered on the server. So, now that `data-reactroot` is unreliable, I replaced that check with `!!domNode.innerHTML` so the hydrate check is now:

```js
const shouldHydrate = !!ReactDOM.hydrate && !!domNode.innerHTML;
```

I tested this most recent commit locally on a page using React 16 and a combination of server/client rendered components. Here is the output from the console:

```
(index):68 [SERVER] RENDERED SearchForm to dom node with id: SearchForm-react-component with railsContext: {...,"serverSide":true}
(index):98 [SERVER] RENDERED Explore to dom node with id: Explore-react-component with railsContext: {...,"serverSide":true}
createReactElement.js:41 HYDRATED SearchForm in dom node with id: SearchForm-react-component using props, railsContext: {} {…,"serverSide":false}
createReactElement.js:41 HYDRATED Explore in dom node with id: Explore-react-component using props, railsContext: {} {…,"serverSide":false}
createReactElement.js:43 RENDERED FreeReport to dom node with id: FreeReport-react-component with props, railsContext: {} {…,"serverSide":false}
createReactElement.js:43 RENDERED ClientOnlyTest to dom node with id: ClientOnlyTest-react-component with props, railsContext: {} {…,"serverSide":false}
```

justin808 commented:
@alexfedoseev Please look at the latest change. I know we want to use this as well.

justin808 commented:
@theJoeBiz can you please rebase on top of master and add an entry to the CHANGELOG.md

theJoeBiz commented:
@justin808 added an entry to CHANGELOG.md and squashed the commits

alexfedoseev commented:
LGTM

justin808 commented:
Shipped! 10.1.2

theJoeBiz commented:
Any estimate on getting this onto npm as 10.1.2?
